### PR TITLE
UTY-1112: Fix List, Map, Option Serialization/Deserialization

### PR DESCRIPTION
**Pull request by jamiebrynes7**
#### Description
Previously had a bug where serialization and deserialization of empty lists, maps, and options was broken as there was no way to distinguish between an option<empty> and option<list<empty>>. Got a patch for the Worker SDK to expose the required APIs and fixed the code.

**READ IN COMMIT ORDER** - last commit is generated code.
#### Tests
Ran Playground with the changes and the workaround removed ✔️ 
Ran Test Project and ensured exhaustive schema compiled ✔️ 
#### Documentation
Will touch documentation in a sibling PR.
#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/src/Generation/Model/Details/UnityFieldDetails.cs [hunk 1]
```diff
@@ -11,6 +11,7 @@ public class UnityFieldDetails
         public string CamelCaseName => Formatting.SnakeCaseToCamelCase(rawFieldDefinition.name);
         public int FieldNumber => rawFieldDefinition.Number;
         public bool IsBlittable;
+        public bool IsListMapOrOption;
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/src/Generation/Model/Details/UnityFieldDetails.cs]: `CanBeEmpty`?

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/src/Generation/Model/Details/UnityFieldDetails.cs]: I like that much better

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/PositionSendSystem.cs [hunk 2]
```diff
@@ -62,7 +62,7 @@ protected override void OnUpdate()
                 var entityId = positionData.SpatialEntityIds[i].EntityId;
 
                 var update = new SchemaComponentUpdate(position.ComponentId);
-                Position.Serialization.Serialize(position, update.GetFields());
+                Position.Serialization.SerializeUpdate(position, update);
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/PositionSendSystem.cs]: We should really make helper functions for this. There is no reason a user has to touch `Serialization`.

Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/PositionSendSystem.cs]: Not for this PR though.

File: workers/unity/Assets/Playground/Scripts/MonoBehaviours/DeleteCubeCommandReceiver.cs [hunk 3]
```diff
@@ -71,7 +71,7 @@ private void OnDeleteEntityResponse(WorldCommands.DeleteEntity.ReceivedResponse
             }
 
             var spawnedCubesCopy =
-                new List<EntityId>(CubeSpawnerInputBehaviour.GetSpawnedCubes(cubeSpawnerWriter.Data));
+                new List<EntityId>(cubeSpawnerWriter.Data.SpawnedCubes);
```
Comment (samiwh) [inline workers/unity/Assets/Playground/Scripts/MonoBehaviours/DeleteCubeCommandReceiver.cs]: We should just remove this. I keep commenting on this code that removing it from that list is wrong.

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/PositionSendSystem.cs (hunk 2 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/PositionSendSystem.cs]: Sort of like a:

`ComponentUpdate CreateComponentUpdate(Position.Component component)` sort of thing?

File: workers/unity/Assets/Playground/Scripts/MonoBehaviours/DeleteCubeCommandReceiver.cs (hunk 3 above)
Comment (jamiebrynes7) [reply workers/unity/Assets/Playground/Scripts/MonoBehaviours/DeleteCubeCommandReceiver.cs]: Maybe so - but out of scope for this PR imp

Comment (jamiebrynes7) [reply workers/unity/Assets/Playground/Scripts/MonoBehaviours/DeleteCubeCommandReceiver.cs]: Its large enough already 😬

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/PositionSendSystem.cs (hunk 2 above)
Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/PositionSendSystem.cs]: Yes. Could take it a step further and make a wrapper in the components send system that also sends it in the connection. Ideally users never need to touch that in these classes too. protected methods > protected fields.

File: workers/unity/Assets/Playground/Scripts/MonoBehaviours/DeleteCubeCommandReceiver.cs (hunk 3 above)
Comment (samiwh) [reply workers/unity/Assets/Playground/Scripts/MonoBehaviours/DeleteCubeCommandReceiver.cs]: Fair enough

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/src/Generation/Model/Details/UnityFieldDetails.cs (hunk 1 above)
Comment (zeroZshadow) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/src/Generation/Model/Details/UnityFieldDetails.cs]: Strings can also be empty however

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/src/Generation/Model/Details/UnityFieldDetails.cs]: `CanBeEmptyCollection` 😉

**Pull request merged**